### PR TITLE
Fix mistaken detection of atomic_container when rpm-ostree is present (fixes #74578)

### LINKED ISSUE #74578: ansible_pkg_mgr fact always returns atomic_container if rpm-ostree is present (breaks package module on some systems)
### Summary

For some modules, notably the `package` module, Ansible relies on the `ansible_pkg_mgr" fact to decide which package manager it should be using.

On systems where `rpm-ostree` is present at `/usr/bin/rpm-ostree`, Ansible makes the assumption that the system's dominant package manager is `rpm-ostree` and returns `atomic_container` as the entry for the `ansible_pkg_mgr` fact.

This isn't always a safe assumption to make, however. On systems with [OSBuild](https://www.osbuild.org/) installed, particularly OSBuild Composer, there's a dependency tree that leads to `rpm-ostree` being pulled in.

You can see this by installing the RHEL 8 / CentOS 8 / Rocky 8 package `osbuild-composer`. It has a dependency on `osbuild-ostree`, which in turn depends on `rpm-ostree`.

So I have quite a few (standard, `dnf`-based) RHEL and Rocky systems which do have `/usr/bin/rpm-ostree` present for legitimate reasons, but not as the dominant package manager.

This is reproducible in 2.9.20 and 2.10.9. I haven't tried other versions but the problem is still present in the code in the devel branch.

### Issue Type

Bug Report

### Component Name

pkg_mgr.py

### Ansible Version

```console
╰─ ansible --version 
ansible 2.10.9
  config file = None
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, Apr 12 2021, 07:42:28) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
```

### Configuration

```console
╰─ ansible-config dump --only-changed
ANSIBLE_NOCOWS(env: ANSIBLE_NOCOWS) = True
```

### OS / Environment

RHEL 8, Rocky 8

### Steps to Reproduce

```sh
╰─ ansible -m setup localhost | grep pkg
        "ansible_pkg_mgr": "dnf",
```

```sh
╰─ sudo dnf install osbuild-composer -y
[...]
(or any package that depends on rpm-ostree)
[...]
```

```sh
╰─ ansible -m setup localhost | grep pkg
        "ansible_pkg_mgr": "atomic_container",
```

### Expected Results

The ansible_pkg_mgr fact should not change just because rpm-ostree is installed. It should only change if the host's actual package manager is ostree.

### Actual Results

```console
The ansible_pkg_mgr fact changes just because rpm-ostree is installed.
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by alexhaydock**
##### SUMMARY
Fixes #74578

For some modules, notably the package module, Ansible relies on the `ansible_pkg_mgr" fact to decide which package manager it should be using.

On systems where rpm-ostree is present at /usr/bin/rpm-ostree, Ansible makes the assumption that the system's dominant package manager is rpm-ostree and returns atomic_container as the entry for the ansible_pkg_mgr fact.

This isn't always a safe assumption to make, however. On systems with OSBuild installed, particularly OSBuild Composer, there's a dependency tree that leads to rpm-ostree being pulled in. There are probably other reasons to have it installed that don't involve it being the system's actual package manager.

This PR addresses the problem by moving `atomic_container` to the lowest priority in the list of dicts that `pkg_mgr.py` uses to evaluate the system's dominant package manager. I have moved it to the lowest priority, below `yum` and `dnf` because this is a problem that is primarily going to affect `rpm` based distributions (since it involves `rpm-ostree` being wrongly detected).

It's fairly safe to move this to the lowest point on the list because we now have #73445 merged, which becomes the first check that runs inside `class PkgMgrFactCollector(BaseFactCollector)`.

See below for the current code in `devel`:
```py
    def _check_rh_versions(self, pkg_mgr_name, collected_facts):
        if os.path.exists('/run/ostree-booted'):
            return "atomic_container"
```

With the above in mind, we already have a much more reliable method of determining whether a system is _actually_ using rpm-ostree as its dominant package manager running as the first step in the pkg_mgr detection process. For this reason, I don't think we risk breaking anything by moving atomic_container down to the bottom of the priority list.

Either way, it has to be lower priority than `yum` or `dnf` to properly resolve #74578, so we either need to de-prioritise `rpm-ostree` or bump up the priority of `yum` and `dnf`. I think this is the preferred approach.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
ansible_pkg_mgr fact (via pkg_mgr.py)

##### ADDITIONAL INFORMATION
Please see the bug description in #74578

### CODE REVIEW

File: lib/ansible/module_utils/facts/system/pkg_mgr.py
Comment (Akasurde) [inline lib/ansible/module_utils/facts/system/pkg_mgr.py:16]: Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) and test for this change. See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

Comment (alexhaydock) [reply lib/ansible/module_utils/facts/system/pkg_mgr.py:16]: Thanks :+1: I have added a change fragment now.

**Review by bcoca — CHANGES_REQUESTED** [review]
You need to verify you don't break the original use case for this https://github.com/ansible/ansible/pull/33721

Also note that we look at `/run/ostree-booted` to set this as the pkg mgr.

**Comment by alexhaydock** [comment]
> You need to verify you don't break the original use case for this #33721

This change shouldn't break the original use case described in #33721 mainly because of what you mention here:

> Also note that we look at `/run/ostree-booted` to set this as the pkg mgr.

This method of checking for ostree-based systems was introduced in #73445 and is much more reliable to determine the package manager that the OS is actually booted with than simply looking for the presence of the `rpm-ostree` binary.

The logic here is that there are many reasons why you might have a `yum` or `dnf` distribution with the `rpm-ostree` binary present too, but you should never realistically see a situation where you have a `yum` or `dnf` distribution with `/run/ostree-booted` present.

Even if there's a ostree-based system with `yum` or `dnf` present (even just as a symlink, or a wrapper which advises the user to use `rpm-ostree` instead), the fact that #73445 has been merged now means my change shouldn't break the `pkg_mgr` fact for those systems.

**Comment by bcoca** [comment]
Yes, that is the assumption, I'm just asking that we verify that before this is merged.

**Comment by alexhaydock** [comment]
> Yes, that is the assumption, I'm just asking that we verify that before this is merged.

Fedora Atomic Host is no longer maintained but I could test this on Fedora Silverblue to ensure it's not a breaking change.

**Comment by alexhaydock** [comment]
So far the theory is holding up on CentOS 8 (i've removed the `devel` branch warnings from the output below):

### Install devel build
```sh
[root@3bf35f8895b6 host]# python3 -m pip install ansible-core-2.12.0.dev0.tar.gz 
```

### Check pkg_mgr fact
```sh
[root@3bf35f8895b6 host]# ansible -m setup localhost | grep pkg
        "ansible_pkg_mgr": "dnf",
```

### Install `rpm-ostree`
```sh
[root@3bf35f8895b6 host]# dnf install rpm-ostree
```

### Check pkg_mgr fact again
```sh
[root@3bf35f8895b6 host]# ansible -m setup localhost | grep pkg
        "ansible_pkg_mgr": "dnf",
```

### Simulate being an ostree-booted system
```sh
[root@3bf35f8895b6 host]# touch /run/ostree-booted
```

### Check pkg_mgr fact once more
```sh
[root@3bf35f8895b6 host]# ansible -m setup localhost | grep pkg
        "ansible_pkg_mgr": "atomic_container",
```

I'll do some proper testing on Fedora Silverblue and I can try and take a look at CoreOS too. With the above in mind the only reason I can see this failing is if those systems don't reliably have the `/run/ostree-booted` file present, but I think they will.

Will do some testing and report back.

**Comment by alexhaydock** [comment]
Testing on Fedora Silverblue 33 shows that this works as expected:
```sh
[user@silverblue33 ~]$ ansible --version
ansible [core 2.12.0.dev0] 
  config file = None
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/.local/lib/python3.9/site-packages/ansible
  ansible collection location = /home/user/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/user/.local/bin/ansible
  python version = 3.9.0 (default, Oct  6 2020, 00:00:00) [GCC 10.2.1 20200826 (Red Hat 10.2.1-3)]
  jinja version = 2.11.3
  libyaml = True

[user@silverblue33 ~]$ ansible -m setup localhost | grep pkg
        "ansible_pkg_mgr": "atomic_container",
```

This system reliably presents the /run/ostree-booted file and does not have yum or dnf present:
```sh
[user@silverblue33 ~]$ ls -lah /run/ostree-booted 
-rw-r-----. 1 root root 0 May  7 10:17 /run/ostree-booted

[user@silverblue33 ~]$ which yum
/usr/bin/which: no yum in (/home/user/.local/bin:/home/user/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin)

[user@silverblue33 ~]$ which dnf
/usr/bin/which: no dnf in (/home/user/.local/bin:/home/user/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin)
```

As a bonus, after installing dnf on the rpm-ostree system just to make sure I haven't created the same problem but in reverse:
```sh
[user@silverblue33 ~]$ which dnf
/usr/bin/dnf

[user@silverblue33 ~]$ ansible -m setup localhost | grep pkg
        "ansible_pkg_mgr": "atomic_container",
```

I think this is probably good to go without risking the use case in #33721.

**Review by bcoca — APPROVED** [review]

**Comment by bcoca** [comment]
@alexhaydock I had meant for someone on our team to do the testing, in any case, thanks for the extra effort!

**Comment by bcoca** [comment]
related #74661

**Pull request merged**
